src/vm/object_model.rs

```diff
@@ -48,17 +48,21 @@ pub trait ObjectModel<VM: VMBinding> {
     //
     // MMTk reserved PolicySpecific side metadata offsets:
     //
-    //  1 - MarkSweep Alloc bit:
+    //  1 - MarkSweep Chunk Mark byte:
     //      - Offset `0x0` on 32-bits
     //      - Offset `LOCAL_SIDE_METADATA_BASE_ADDRESS` on 64-bits
+    //  2 - MarkSweep Alloc bit:
+    //      - Offset `Chunk Mark`.offset + `Chunk Mark`.metadata_address_range_size()
+    //  3 - MarkSweep Page Mark byte:
+    //      - Offset `Alloc bit`.offset + `Alloc bit`.metadata_address_range_size()
```

Comment:
For the side metadata specs here, you would need to update: https://github.com/mmtk/mmtk-core/blob/a8499062f9df749a6cf7acf1e626d642913206bb/src/util/metadata/side_metadata/constants.rs#L61

I am doing a PR, and moving these comments to the constants. We may need a better way to deal with this to make sure when we use new side metadata, the base address is updated properly. But now, we have to manually update it.

Reply:
@k-sareen

Reply:
Oh it seems you have done it.

Reply:
I should have done it yeah. This is something I did want to bring up because it's currently a bit obtuse and maybe we need to replace it with something like a `lazy_static!{}` that can do this during compilation for us

Reply:
Those needs to be constants so `lazy_static!` won't work. We may need either macros or custom build scripts.

Reply:
Oh of course yes, my bad. Hmm yeah macros could work

Reply:
Would it not just be easier to pick a relatively safe base address for the VM metadata instead? Though there is still a chance for overlap I guess

Reply:
There are assertions to make sure side metadata specs don't overlap. I don't know if it is a hard requirement that the side metadata specs need to be placed adjacent to each other. @javadamiri But I think as long as we have clear documents on how to define side metadata specs for bindings, and have assertions to prevent invalid specs, we are okay. We just need to improve the process so it is easier and less error prone to add new specs.

Reply:
To be honest I am still a bit confused regarding how the 32-bit metadata is supposed to work. I think that might be the cause for the JikesRVM crash (I haven't looked into it yet). I'm always confused where to use `metadata_address_range_size()` and where to use `metadata_bytes_per_chunk()`

Reply:
It's not a requirement.
The pros of doing this are:
1 - In chunked side_metadata, the adjacency saves us some memory space.
2 - In contiguous side_metadata, it may save us some address range.

Otherwise, as long as the spec offsets are constants, there shouldn't be any other concerns

Reply:
`metadata_address_range_size()` is used for contiguous side metadata and it calculates the whole address range size.
We use it for:
1 - global and local side metadata in 64-bits,
2 - global side metadata in 32-bits.

`metadata_bytes_per_chunk()` is used for chunked metadata, and calculates the side metadata space size required for 1 chunk (`2 ^ 22` bytes).
We use it for local side metadata in 32-bits.

The reason why we considered chunked side metadata is that in 32-bits targets the address range size is limited and we won't be able to provide enough address space for all local side metadata.
By making it chunked, we use the fact that each chunk is only managed by one policy, and only the local side metadata of that policy will be sufficient.

Reply:
Ah so then I should conditionally compile the offset for the local side metadata here?

https://github.com/mmtk/mmtk-core/pull/361/files#diff-cdc1b56966139dcff8d3cc22c88746f0d16767b5e27ef37fb81bd5b08e5f375aR49

Reply:
Yes, based on `#[cfg(target_pointer_width = "32")]` or `#[cfg(target_pointer_width = "64")]`.

Reply:
I'll update that -- it's likely the JikesRVM issue then